pip/pip-321.md

```diff
@@ -0,0 +1,311 @@
+
+**Explanation for 409 Conflict:** This follows the behavior of namespace replication clusters. As per PIP-8, a peer-cluster cannot be part of a replication-cluster. Similarly, for allowed-clusters, users could enable replication at the topic level, hence a peer-cluster cannot be part of allowed-clusters as well.
+
+#### `getNamespaceAllowedClusters` Endpoint
+
+This new endpoint allows retrieving the list of allowed clusters for a specific namespace.
+
+**Method:**
+```
+GET
+```
+
+**Path:**
+```
+/namespaces/{tenant}/{namespace}/allowedClusters
+```
+
+**Response Codes:**
+
+- `403 Forbidden`: The requester does not have admin permissions.
+- `404 Not Found`: The specified tenant, cluster, or namespace does not exist.
+- `412 Precondition Failed`: The namespace is not global.
+
+**Example Response:**
+```
+[
+    "cluster1",
+    "cluster2",
+    "cluster3"
+]
+```
+
+### Binary protocol
+
+### Configuration
+
+### CLI
+
+#### `setNamespaceAllowedClusters` Command
+
+This new command allows you to set the list of allowed clusters for a specific namespace.
+
+**Usage:**
+
+```
+$ pulsar admin namespaces set-allowed-clusters --clusters <cluster-ids> <tenant>/<namespace>
+```
+
+**Options:**
+   * --clusters, -c
+     - A comma-separated list of cluster IDs.
+
+**Response Codes:**
+
+- `400 Bad Request`: The allowed clusters should contain all replication clusters.
+- `403 Forbidden`: You do not have admin permission.
+- `404 Not Found`: The tenant, cluster, or namespace does not exist.
+- `409 Conflict`: A peer-cluster cannot be part of an allowed-cluster.
+- `412 Precondition Failed`: The namespace is not global or the cluster IDs are invalid.
+
+**Explanation for 409 Conflict:** This follows the behavior of namespace replication clusters. In PIP-8, it introduced the concept of a peer-cluster, which cannot be part of a replication-cluster. For the allowed-clusters, users could enable replication at the topic level, so the peer-cluster cannot be part of the allowed-clusters too.
+
+#### `getNamespaceAllowedClusters` Command
+
+This new command allows you to retrieve the list of allowed clusters for a specific namespace.
+
+**Usage:**
+
+```
+$ pulsar admin namespaces get-allowed-clusters <tenant>/<namespace>
+```
+
+**Response Codes:**
+
+- `403 Forbidden`: You do not have admin permission.
+- `404 Not Found`: The tenant, cluster, or namespace does not exist.
+- `412 Precondition Failed`: The namespace is not global.
+
+**Example Response:**
+
+```
+"cluster1"
+"cluster2"
+"cluster3"
+```
+
+#### `CreateNamespace` Command
+
+Add a new option to this command to set allowed clusters when create a new namespace.
+
+**Usage:**
+
+```
+$ pulsar admin namespaces create [options] tenant/namespace
+```
+
+**Options:**
+   * --bundles, -b
+      - number of bundles to activate. Default: 0
+   * --clusters, -c 
+     - List of replication clusters this namespace will be assigned. (Modified*)
+   * --allowed-clusters, -a (New*)
```

Comment:
Please mention that if `--allowed-clusters` is absent. The value of `--clusters` will be used as the allowed clusters.
And do you need to change the REST API to support setting allowed clusters when creating a namespace?

Reply:
>Please mention that if --allowed-clusters is absent. The value of --clusters will be used as the allowed clusters.
Good point. Thanks for the reminder.

Reply:
>And do you need to change the REST API to support setting allowed clusters when creating a namespace?

I replay it in https://github.com/apache/pulsar/pull/21648#discussion_r1461997483.
I think there are two approaches:
1. Mark `void createNamespace(String namespace, Set<String> clusters)` as deprecated.
2. Rename method param name of `void createNamespace(String namespace, Set<String> clusters)`. And add two method `createNamespace(String namespace, Set<String> allowedClusters)` and `createNamespace(String namespace, Set<String> replicationClusters, Set<String> allowedClusters)`.

Reply:
I see. We have 

```
    @PUT
    @Path("/{tenant}/{namespace}")
    @ApiOperation(value = "Creates a new namespace with the specified policies")
    @ApiResponses(value = { @ApiResponse(code = 403, message = "Don't have admin permission"),
            @ApiResponse(code = 404, message = "Tenant or cluster doesn't exist"),
            @ApiResponse(code = 409, message = "Namespace already exists"),
            @ApiResponse(code = 412, message = "Namespace name is not valid") })
    public void createNamespace(@Suspended AsyncResponse response,
                                @PathParam("tenant") String tenant,
                                @PathParam("namespace") String namespace,
                                @ApiParam(value = "Policies for the namespace") Policies policies)
```

So we don't need to change the method signature.